File: backend/routes/suggestions.py
import os
import requests
from flask import Blueprint, request, jsonify
from firebase_admin import db
from .auth import verify_token  # Your existing token verification function
from datetime import datetime

suggestions_bp = Blueprint('suggestions', __name__)
GROQ_API_KEY = os.environ.get('GROQ_API_KEY')

def get_current_season():
    month = datetime.now().month
    if month in [12, 1, 2]:
        return 'winter'
    elif month in [3, 4, 5]:
        return 'spring'
    elif month in [6, 7, 8]:
        return 'summer'
    else:
        return 'fall'

def get_groq_suggestions(prompt: str) -> str:
    url = 'https://api.groq.com/openai/v1/chat/completions'
    headers = {
        'Authorization': f'Bearer {GROQ_API_KEY}',
        'Content-Type': 'application/json'
    }
    data = {
        'model': 'llama-3.3-70b-versatile',
        'messages': [
            {'role': 'system', 'content': (
                "You are a helpful assistant suggesting grocery items. "
                "Include seasonal and trending products, categorize items, "
                "and suggest quantities as needed."
            )},
            {'role': 'user', 'content': prompt}
        ],
        'max_tokens': 200
    }
    response = requests.post(url, headers=headers, json=data)
    if response.ok:
        return response.json()['choices'][0]['message']['content']
    else:
        print('Groq API error:', response.status_code, response.text)
        return f'Error: {response.status_code} {response.text}'

@suggestions_bp.route('/suggestions', methods=['POST'])
def suggestions():
    token = request.headers.get('Authorization')
    if token and token.startswith('Bearer '):
        token = token[7:]
    else:
        return jsonify({'error': 'Authorization token required'}), 401

    uid = verify_token(token)
    if not uid:
        return jsonify({'error': 'Invalid token'}), 401

    # Get current season
    season = get_current_season()

    # Fetch user's cart items from Firebase
    cart_ref = db.reference(f'carts/{uid}')
    cart = cart_ref.get() or {}

    # Prepare cart summary string
    cart_items_list = [f"{item.get('quantity',1)} x {item.get('name', 'unknown')}" for item in cart.values()]
    cart_summary = ", ".join(cart_items_list) if cart_items_list else "no items"

    # Use the incoming prompt from the user (optional, or you can build your own)
    data = request.get_json(silent=True) or {}
    user_prompt = data.get('prompt', '').strip() or "Suggest grocery items"

    # Construct prompt with cart and season context
    prompt = (
        f"The user currently has {cart_summary} in their cart. "
        f"It is {season} season. "
        f"Based on this, {user_prompt}."
    )

    suggestions_text = get_groq_suggestions(prompt)
    return jsonify({'suggestions': suggestions_text})

File: backend-go/handlers/suggestions.go
package handlers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"

	"echobasket/config"
	"echobasket/firebase"
)

type SuggestionsRequest struct {
	Prompt string `json:"prompt"`
}

// getCurrentSeason returns the current season based on month
func getCurrentSeason() string {
	month := time.Now().Month()
	switch {
	case month >= 12 || month <= 2:
		return "winter"
	case month >= 3 && month <= 5:
		return "spring"
	case month >= 6 && month <= 8:
		return "summer"
	default:
		return "fall"
	}
}

// getGroqSuggestions calls the Groq API for AI suggestions
func getGroqSuggestions(prompt string) (string, error) {
	url := "https://api.groq.com/openai/v1/chat/completions"

	payload := map[string]interface{}{
		"model": "llama-3.3-70b-versatile",
		"messages": []map[string]string{
			{
				"role": "system",
				"content": "You are a helpful assistant suggesting grocery items. " +
					"Include seasonal and trending products, categorize items, " +
					"and suggest quantities as needed.",
			},
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"max_tokens": 200,
	}

	jsonData, _ := json.Marshal(payload)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}

	req.Header.Set("Authorization", "Bearer "+config.AppConfig.GroqAPIKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("Groq API error: %d %s", resp.StatusCode, string(body))
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", err
	}

	choices := result["choices"].([]interface{})
	if len(choices) > 0 {
		message := choices[0].(map[string]interface{})["message"].(map[string]interface{})
		return message["content"].(string), nil
	}

	return "", fmt.Errorf("no response from Groq")
}

// Suggestions handles the /suggestions endpoint
func Suggestions(c *gin.Context) {
	uid := defaultUID

	season := getCurrentSeason()

	// Fetch user's cart items
	ctx := context.Background()
	ref := firebase.DBClient.NewRef("carts/" + uid)

	var cart map[string]map[string]interface{}
	if err := ref.Get(ctx, &cart); err != nil {
		cart = make(map[string]map[string]interface{})
	}

	// Prepare cart summary
	var cartItems []string
	for _, item := range cart {
		quantity := 1
		if q, ok := item["quantity"].(float64); ok {
			quantity = int(q)
		}
		name := "unknown"
		if n, ok := item["name"].(string); ok {
			name = n
		}
		cartItems = append(cartItems, fmt.Sprintf("%d x %s", quantity, name))
	}

	cartSummary := "no items"
	if len(cartItems) > 0 {
		cartSummary = strings.Join(cartItems, ", ")
	}

	// Get user prompt
	var req SuggestionsRequest
	c.ShouldBindJSON(&req)
	userPrompt := req.Prompt
	if userPrompt == "" {
		userPrompt = "Suggest grocery items"
	}

	// Construct prompt with cart and season context
	prompt := fmt.Sprintf(
		"The user currently has %s in their cart. It is %s season. Based on this, %s.",
		cartSummary, season, userPrompt,
	)

	suggestionsText, err := getGroqSuggestions(prompt)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"suggestions": suggestionsText})
}
